subprojects/core/src/main/java/org/gradle/caching/configuration/BuildCacheConfiguration.java

```diff
@@ -43,17 +44,16 @@
  * @since 3.5
  */
 @Incubating
+@HasInternalProtocol
 public interface BuildCacheConfiguration {
 
     /**
-     * Registers an instance of a {@link BuildCacheServiceFactory} that can create a
-     * {@link org.gradle.caching.BuildCacheService} given a {@link BuildCache} configuration.
+     * Registers a custom build cache type.
      *
-     * A factory must be registered before a {@code BuildCacheService} instance can be created.
-     *
-     * @param buildCacheServiceFactory instance to register
+     * @param configurationType Configuration object used to provide parameters to a {@link BuildCacheService}
+     * @param buildCacheServiceType Implementation of {@link BuildCacheService} that should be created when this build cache type is used
      */
-    void registerBuildCacheServiceFactory(BuildCacheServiceFactory<?> buildCacheServiceFactory);
+    void registerBuildCacheService(Class<? extends BuildCache> configurationType, Class<? extends BuildCacheService> buildCacheServiceType);
```

Comment:
We only use the registration for our internal, built-in build cache services so we don't have to apply a plugin.  We don't need to expose the registration as a public type or make it easier since that set will be fixed or it will be an external plugin (and use the regular API).